### PR TITLE
[Support] Vendor rpmalloc in-tree and use it for the Windows 64-bit release

### PR DESCRIPTION
**Pull request by aganea**
### Context

We have a longstanding performance issue on Windows, where to this day, the default heap allocator is still lockfull. With the number of cores increasing, building and using LLVM with the default Windows heap allocator is sub-optimal. Notably, the ThinLTO link times with LLD are extremely long, and increase proportionally with the number of cores in the machine.

In https://github.com/llvm/llvm-project/commit/a6a37a2fcd2a8048a75bd0d8280497ed89d73224, I introduced the ability build LLVM with several popular lock-free allocators. Downstream users however have to build their own toolchain with this option, and building an optimal toolchain is a bit tedious and long. Additionally, LLVM is now integrated into Visual Studio, which AFAIK re-distributes the vanilla LLVM binaries/installer. The point being that many users are impacted and might not be aware of this problem, or are unable to build a more optimal version of the toolchain.

The symptom before this PR is that most of the CPU time goes to the kernel (darker blue) when linking with ThinLTO:

16c_ryzen9_windows_heap

With this PR, most time is spent in user space (light blue):

16c_ryzen9_rpmalloc

On higher core count machines, before this PR, the CPU usage becomes pretty much flat because of contention:

VM_176_windows_heap

With this PR, similarily most CPU time is now used:

VM_176_with_rpmalloc

### Changes in this PR

The avenue I've taken here is to vendor/re-licence rpmalloc in-tree, and use it when building the Windows 64-bit release. Given the permissive rpmalloc licence, prior discussions with the LLVM foundation and @lattner suggested this vendoring. Rpmalloc's author (@mjansson) kindly agreed to ~~donate~~ re-licence the rpmalloc code in LLVM (please do correct me if I misinterpreted our past communications).

I've chosen rpmalloc because it's small and gives the best value overall. The source code is only 4 .c files. Rpmalloc is statically replacing the weak CRT alloc symbols at link time, and has no dynamic patching like mimalloc. As an alternative, there were several unsuccessfull attempts made by Russell Gallop to use SCUDO in the past, please see thread in https://reviews.llvm.org/D86694. If later someone comes up with a PR of similar performance that uses SCUDO, we could then delete this vendored rpmalloc folder.

I've added a new cmake flag `LLVM_ENABLE_RPMALLOC` which essentialy sets `LLVM_INTEGRATED_CRT_ALLOC` to the in-tree rpmalloc source.

### Performance

The most obvious test is profling a ThinLTO linking step with LLD. I've used a Clang compilation as a testbed, ie.
```
set OPTS=/GS- /D_ITERATOR_DEBUG_LEVEL=0 -Xclang -O3 -fstrict-aliasing -march=native -flto=thin -fwhole-program-vtables -fuse-ld=lld
cmake -G Ninja %ROOT%/llvm -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_ASSERTIONS=TRUE -DLLVM_ENABLE_PROJECTS="clang" -DLLVM_ENABLE_PDB=ON -DLLVM_OPTIMIZED_TABLEGEN=ON -DCMAKE_C_COMPILER=clang-cl.exe -DCMAKE_CXX_COMPILER=clang-cl.exe -DCMAKE_LINKER=lld-link.exe -DLLVM_ENABLE_LLD=ON -DCMAKE_CXX_FLAGS="%OPTS%" -DCMAKE_C_FLAGS="%OPTS%" -DLLVM_ENABLE_LTO=THIN
```
I've profiled the linking step with no LTO cache, with Powershell, such as:
```
Measure-Command { lld-link /nologo @CMakeFiles\clang.rsp /out:bin\clang.exe /implib:lib\clang.lib /pdb:bin\clang.pdb /version:0.0 /machine:x64 /STACK:10000000 /DEBUG /OPT:REF /OPT:ICF /INCREMENTAL:NO /subsystem:console /MANIFEST:EMBED,ID=1 }`
```

Timings:

| Machine | Allocator | Time to link |
|--------|--------|--------|
| 16c/32t AMD Ryzen 9 5950X | Windows Heap | 10 min 38 sec |
|  | **Rpmalloc** | **4 min 11 sec** |
| 32c/64t AMD Ryzen Threadripper PRO 3975WX | Windows Heap | 23 min 29 sec |
|  | **Rpmalloc** | **2 min 11 sec** |
|  | **Rpmalloc + /threads:64** | **1 min 50 sec** |
| 176 vCPU (2 socket) Intel Xeon Platinum 8481C (fixed clock 2.7 GHz) | Windows Heap | 43 min 40 sec |
|  | **Rpmalloc** | **1 min 45 sec** |

This also improves the overall performance when building with clang-cl. I've profiled a regular compilation of clang itself, ie:
```
set OPTS=/GS- /D_ITERATOR_DEBUG_LEVEL=0 /arch:AVX -fuse-ld=lld
cmake -G Ninja %ROOT%/llvm -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_ASSERTIONS=TRUE -DLLVM_ENABLE_PROJECTS="clang;lld" -DLLVM_ENABLE_PDB=ON -DLLVM_OPTIMIZED_TABLEGEN=ON -DCMAKE_C_COMPILER=clang-cl.exe -DCMAKE_CXX_COMPILER=clang-cl.exe -DCMAKE_LINKER=lld-link.exe -DLLVM_ENABLE_LLD=ON -DCMAKE_CXX_FLAGS="%OPTS%" -DCMAKE_C_FLAGS="%OPTS%"
```
This saves approx. 30 sec when building on the Threadripper PRO 3975WX:
```
(default Windows Heap)
C:\src\git\llvm-project>hyperfine -r 5 -p "make_llvm.bat stage1_test2" "ninja clang -C stage1_test2"
Benchmark 1: ninja clang -C stage1_test2
  Time (mean ± σ):     392.716 s ±  3.830 s    [User: 17734.025 s, System: 1078.674 s]
  Range (min … max):   390.127 s … 399.449 s    5 runs

(rpmalloc)
C:\src\git\llvm-project>hyperfine -r 5 -p "make_llvm.bat stage1_test2" "ninja clang -C stage1_test2"
Benchmark 1: ninja clang -C stage1_test2
  Time (mean ± σ):     360.824 s ±  1.162 s    [User: 15148.637 s, System: 905.175 s]
  Range (min … max):   359.208 s … 362.288 s    5 runs
```

### CODE REVIEW

**Comment by llvm-beanz** [comment]
@mjansson, as the original author of much of this code can you please confirm on this thread that you are consenting to it being contributed under the LLVM licensing terms? The current licensing terms are documented in the developer policy here: https://github.com/llvm/llvm-project/blob/main/llvm/docs/DeveloperPolicy.rst#copyright-license-and-patents.

At the moment this requires all code to be licensed under both the new Apache w/ LLVM license and the legacy NCSA license.

There is a plan to only require the new license in the future (see: https://discourse.llvm.org/t/relicensing-next-step-dropping-requirement-to-contribute-also-under-the-legacy-license/78351), but for now all code contributed must be available under either license.

Thank you!

**Comment by mjansson** [comment]
As author of rpmalloc I consent to it being contributed under the LLVM licensing terms, both the new Apache w/ LLVM license and the legacy NCSA license.

**Comment by zmodem** [comment]
Sounds great to me. Getting this into the windows packages without having to jump through hoops will be very nice (and I look forward to removing the hoop jumping we do in Chromium's clang build script).

I assume this is picking up the latest rpmalloc release (1.4.5)? We should document somewhere what exact version is being used.

Is there a plan for keeping llvm's rpmalloc copy up-to-date?

Do we want to keep the general `LLVM_INTEGRATED_CRT_ALLOC` support going forward?

**Comment by aganea** [comment]
> I assume this is picking up the latest rpmalloc release (1.4.5)? We should document somewhere what exact version is being used.
I've upgraded to 1.4.5 and added the version number information in README.md

> Is there a plan for keeping llvm's rpmalloc copy up-to-date?
I would assume after the initial commit, upgrades would be as regular contributions?

> Do we want to keep the general `LLVM_INTEGRATED_CRT_ALLOC` support going forward?
I would keep it for now for testing other allocators. I still find mimalloc being a very good contender and I hope it'll be eventually integrated into Windows through some application manifest. It'd be good too if someone with SCUDO experience would make it work `LLVM_INTEGRATED_CRT_ALLOC`.

**Comment by tru** [comment]
> I would keep it for now for testing other allocators. I still find mimalloc being a very good contender and I hope it'll be eventually integrated into Windows through the application manifest maybe. Also, it'd be good if someone with SCUDO experience would make it work with `LLVM_INTEGRATED_CRT_ALLOC`.

If we keep it around - maybe we need to update the documentation to point out that using RPMALLOC is the most supported way and that this option is only for testing other allocators? Seems like we want people to use the vendored version of rpmalloc most of the time.

**Comment by aganea** [comment]
> > I would keep it for now for testing other allocators. I still find mimalloc being a very good contender and I hope it'll be eventually integrated into Windows through the application manifest maybe. Also, it'd be good if someone with SCUDO experience would make it work with `LLVM_INTEGRATED_CRT_ALLOC`.
> 
> If we keep it around - maybe we need to update the documentation to point out that using RPMALLOC is the most supported way and that this option is only for testing other allocators? Seems like we want people to use the vendored version of rpmalloc most of the time.

I've updated the `CMake.rst` file. Can you take a look please see if that suits what you had in mind?

File: llvm/CMakeLists.txt [hunk 1]
```diff
@@ -733,7 +733,24 @@ if( WIN32 AND NOT CYGWIN )
...
+if(LLVM_ENABLE_RPMALLOC)
+  if(NOT (CMAKE_SYSTEM_NAME MATCHES "Windows|Linux"))
+    message(FATAL_ERROR "LLVM_ENABLE_RPMALLOC is only supported on Windows and Linux.")
+  endif()
+  if(LLVM_USE_SANITIZER)
+    message(FATAL_ERROR "LLVM_ENABLE_RPMALLOC cannot be used along with LLVM_USE_SANITIZER!")
+  endif()
+  if(WIN32 AND CMAKE_BUILD_TYPE AND uppercase_CMAKE_BUILD_TYPE STREQUAL "DEBUG")
+    message(FATAL_ERROR "The Debug target isn't supported along with LLVM_ENABLE_RPMALLOC!")
+  endif()
```
Comment (llvm-beanz) [inline llvm/CMakeLists.txt:746]: What happens if you use a multi-configuraiton build generator (i.e. Visual Studio)? In that case this won't trigger but it will generate a debug build configuration.

Comment (aganea) [reply llvm/CMakeLists.txt:757]: As suggested.

**Comment by aganea** [comment]
Just for reference, latest MSVC 17.9.7 is litteraly twice as slow to build LLVM, as compared to this current PR:
```
C:\src\git\llvm-project>hyperfine -r 5 -p "make_llvm.bat stage1_test2" "ninja clang -C stage1_test2"
Benchmark 1: ninja clang -C stage1_test2
  Time (mean ± σ):     784.241 s ± 17.021 s    [User: 28043.391 s, System: 2018.822 s]
  Range (min … max):   761.882 s … 804.272 s    5 runs
```
Some of the LLVM source files seem to be hitting a quadratic behavior somewhere in MSVC: https://developercommunity.visualstudio.com/t/ON2-in-SparseBitVectorBase-when-com/10657991

**Comment by aganea** [comment]
Gentle ping!

What do you all think the way forward is with this patch? Is anything else to be done here? Does this need wider adoption from the community? (as in a RFC)

**Comment by compnerd** [comment]
I think that the legal aspects have been resolved right? At that point, we should consider getting this into the tree so that others can easily play with it. I know that I am interested in exploring this as an alternative memory allocator for the Swift compiler as well.

**Comment by tru** [comment]
Yeah. I think we should merge it, it's off by default so it won't affect people.

**Review by tru — APPROVED** [review]

**Comment by aganea** [comment]
> I think there are still some unresolved comments from my review, but I'm all for getting this merged.

@zmodem Could you please point me specifically to which comments were unresolved? I think I’ve answered here: https://github.com/llvm/llvm-project/pull/91862#issuecomment-2107491507

File: llvm/docs/ReleaseNotes.rst [hunk 2]
```diff
@@ -62,6 +62,9 @@ Changes to LLVM infrastructure
 Changes to building LLVM
 ------------------------
 
+* LLVM now supports rpmalloc in-tree, as a replacement C allocator for hosted
+  toolchains. Currently using version 1.4.5.
```
Comment (zmodem) [inline llvm/docs/ReleaseNotes.rst:66]: Maybe this could be expanded a bit, with a link to the relevant section in CMake.rst, mentioning that the relevant flag is LLVM_ENABLE_RPMALLOC, that the main benefit is for LTO on Windows etc?

File: llvm/lib/Support/rpmalloc/rpmalloc.c [hunk 3]
```diff
@@ -0,0 +1,3993 @@
+//===------------------------ rpmalloc.c -----------------*- C
+//-*-=============//
```
Comment (zmodem) [inline llvm/lib/Support/rpmalloc/rpmalloc.c:2]: Bad linebreak?

File: llvm/lib/Support/rpmalloc/rpmalloc.h [hunk 4]
```diff
@@ -0,0 +1,429 @@
+//===------------------------ rpmalloc.h -----------------*- C
+//-*-=============//
```
Comment (zmodem) [inline llvm/lib/Support/rpmalloc/rpmalloc.h:2]: Bad linebreak?

File: llvm/utils/release/build_llvm_release.bat [hunk 5]
```diff
@@ -163,6 +163,8 @@ set common_cmake_flags=^
   -DCLANG_ENABLE_LIBXML2=OFF ^
   -DCMAKE_C_FLAGS="%common_compiler_flags%" ^
   -DCMAKE_CXX_FLAGS="%common_compiler_flags%" ^
+  -DLLVM_ENABLE_RPMALLOC=ON ^
+  -DLLVM_ENABLE_PDB=ON ^
```
Comment (zmodem) [inline llvm/utils/release/build_llvm_release.bat:167]: Why ENABLE_PDB?

File: llvm/utils/release/build_llvm_release.bat [hunk 6]
```diff
@@ -250,7 +252,7 @@ set cmake_flags=^
   -DLLDB_TEST_COMPILER=%stage0_bin_dir%/clang.exe ^
   -DPYTHON_HOME=%PYTHONHOME% ^
   -DPython3_ROOT_DIR=%PYTHONHOME% ^
-  -DLIBXML2_INCLUDE_DIRS=%libxmldir%/include/libxml2 ^
+  -DLIBXML2_INCLUDE_DIR=%libxmldir%/include/libxml2 ^
```
Comment (zmodem) [inline llvm/utils/release/build_llvm_release.bat:255]: This seems unrelated?

**Review by zmodem — COMMENTED** [review]
> > I think there are still some unresolved comments from my review, but I'm all for getting this merged.
> 
> @zmodem Could you please point me specifically to which comments were unresolved? I think I’ve answered here: [#91862 (comment)](https://github.com/llvm/llvm-project/pull/91862#issuecomment-2107491507)

Sorry, my fault. It seems my comments were still "pending" :(

File: llvm/docs/ReleaseNotes.rst (hunk 2 above)
Comment (aganea) [reply llvm/docs/ReleaseNotes.rst]: As suggested.

File: llvm/utils/release/build_llvm_release.bat (hunk 5 above)
Comment (aganea) [reply llvm/utils/release/build_llvm_release.bat]: Removed, I had it for debugging. Will propose that separately.

File: llvm/utils/release/build_llvm_release.bat (hunk 6 above)
Comment (aganea) [reply llvm/utils/release/build_llvm_release.bat:254]: Removed. We can't build with the latest cmake without this, will propose that in a separate PR.

**Review by zmodem — APPROVED** [review]
The `s/DLIBXML2_INCLUDE_DIRS/DLIBXML2_INCLUDE_DIR/` change is still there after the latest commit. Tackling that in a separate PR sounds good to me.

Besides that, this lgtm

File: llvm/CMakeLists.txt (hunk 1 above)
Comment (jayfoad) [reply llvm/CMakeLists.txt:757]: What's the problem with debug builds? Has this been discussed or described elsewhere?

File: llvm/CMakeLists.txt [hunk 7]
```diff
@@ -733,7 +733,35 @@ if( WIN32 AND NOT CYGWIN )
 endif()
 set(LLVM_NATIVE_TOOL_DIR "" CACHE PATH "Path to a directory containing prebuilt matching native tools (such as llvm-tblgen)")
 
-set(LLVM_INTEGRATED_CRT_ALLOC "" CACHE PATH "Replace the Windows CRT allocator with any of {rpmalloc|mimalloc|snmalloc}. Only works with CMAKE_MSVC_RUNTIME_LIBRARY=MultiThreaded.")
+set(LLVM_ENABLE_RPMALLOC "" CACHE BOOL "Replace the CRT allocator with rpmalloc.")
+if(LLVM_ENABLE_RPMALLOC)
+  if(NOT (CMAKE_SYSTEM_NAME MATCHES "Windows|Linux"))
+    message(FATAL_ERROR "LLVM_ENABLE_RPMALLOC is only supported on Windows and Linux.")
```
Comment (jayfoad) [inline llvm/CMakeLists.txt:739]: Only supported on Windows, surely? Since it relies on `LLVM_INTEGRATED_CRT_ALLOC` which is only supported on Windows (see line 767 below).

**Comment by stefan-il** [comment]
Hi @aganea, just to check my understanding LLVM_ENABLE_RPMALLOC will replace default allocator used on Windows build and will not only help with ThinLTO but also with compilation speed?
Are there any know limitations why this is not also available on Linux, or is it just something that will be covered at later date?

**Comment by aganea** [comment]
@stefan-il Please see discussion in https://reviews.llvm.org/D101427 especially last comment by @MaskRay. We could encode those linker flags for ease of use perhaps, when LLVM_ENABLE_RPMALLOC is used, if someone wants to do it.
